Add diagnostic environment RiskyPath

pedrofreire opened the pull request:
Add simple diagnostic environment testing for correct behavior under stochasticity.

AdamGleave commented on src/seals/diagnostics/__init__.py:
```diff
@@ -0,0 +1,7 @@
+"""Simple diagnostic environments."""
+
+from seals.diagnostics.risky_path import RiskyPathEnv
```
I'd suggest we follow the convention followed in OpenAI Gym which is to:
  - Register all environments in the `__init__.py` file, like in https://github.com/openai/gym/blob/master/gym/envs/__init__.py
  - Do *not* import the submodules defining these environments. This is done lazily when `gym.make` is called.
The advantage of this is (a) submodules can depend on optional dependencies without breaking things if you don't load that environment; (b) it'll keep `import seals` fast.

This isn't a huge deal for these environments since they're all simple Python implementations without external dependencies and which will have minimal overhead, but it's important for other environments we may add in the future which have optional external dependencies like MuJoCo, and I'd rather keep the style consistent throughout.

We should also import `diagnostics` from `src/seals/__init__.py` so that environments are automatically available. This is OK if the `__init__.py` only contains `gym.register` calls which are fast.

AdamGleave commented on src/seals/diagnostics/risky_path.py:
```diff
@@ -0,0 +1,58 @@
+
+_horizon_v0 = 5
+
+gym.register(
```
Move to `__init__.py`

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+from seals.util import sample_distribution
+
+
+class BaseEnv(gym.Env):
```
As I mentioned in our call, we'll likely want to separate out the tabular and non-tabular bits. Don't need to do it in this PR, though, since `RiskyPath` is tabular.

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+from seals.util import sample_distribution
+
+
+class BaseEnv(gym.Env):
```
This overlaps a lot with https://github.com/HumanCompatibleAI/imitation/blob/c8d39a6e524466221935eec4e904dc9254bc2b96/src/imitation/envs/resettable_env.py -- I think we should combine them in some way.

@shwang how would you feel about making `seals` a required dependency of `imitation`, but ensuring that we keep `seals` dependencies light (things like MuJoCo/ALE will be kept optional, like in Gym)? We could then have a `seals.common` that defines some base classes like this.

qxcv replied:
Some abstractions that we could have as mixins:

- Resettable environments
- Transitions/rewards for tabular MDPs
- _Maybe_ some intermediate mixin that gives you an explicit successor distribution for each state? e.g. have a `get_successors(state, action)` function that enumerates all successors of the current state given the current action, along with their probability (or some symbolic representation of that distribution, like a Gaussian). This would be sufficient to implement, e.g., VI, while still being suitable to represent large factored MDPs and continuous MDPs.

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+        """Initialize spaces (if discrete) and select random seed."""
+        super().__init__()
+
+        if num_states is not None:
```
This is confusing: caller doesn't have to define `observation_space` if they specify this but otherwise they do.

Better to make sure mandatory properties are always set. If you split this up into a tabular and non-tabular part it'd be natural; have the non-tabular part take mandatory `observation_space` and `action_space` arguments, and the tabular part take mandatory `num_states` and `num_action` arguments.

AdamGleave commented on src/seals/diagnostics/risky_path.py:
```diff
@@ -0,0 +1,58 @@
+import gym
+import numpy as np
+
+from seals.diagnostics.base_env import BaseEnv
```
```suggestion
from seals.diagnostics import base_env
```

https://google.github.io/styleguide/pyguide.html#22-imports

AdamGleave commented on src/seals/diagnostics/risky_path.py:
```diff
@@ -0,0 +1,58 @@
+from seals.diagnostics.base_env import BaseEnv
+
+
+class RiskyPathEnv(BaseEnv):
```
```suggestion
class RiskyPathEnv(base_env.BaseEnv):
```

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+    def transition_fn(self, state, action):
+        """Sample next state according to T(s, a)."""
+        return sample_distribution(
+            self.transition_matrix[state, action], random=self.np_random,
```
`transition_matrix` has to be defined for this to function, so make it a mandatory argument in `__init__`

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+
+    def reward_fn(self, state, action, new_state):
+        """Return reward value R(s, a, s')."""
+        return self.reward_matrix[state, action, new_state]
```
`reward_matrix` has to be defined for this to function, so make it a mandatory argument in `__init__`

qxcv replied:
Could also make `reward_fn` abstract and split out into a `TabularEnv` mixin?

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+from gym.utils import seeding
+import numpy as np
+
+from seals.util import sample_distribution
```
```suggestion
from seals import util
```

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+"""Minimal base class for simple environments."""
+
+import gym
+from gym.spaces import Discrete
```
```suggestion
from gym import spaces
```

AdamGleave replied:
https://google.github.io/styleguide/pyguide.html#22-imports

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+        super().__init__()
+
+        if num_states is not None:
+            self.observation_space = Discrete(num_states)
```
```suggestion
            self.observation_space = spaces.Discrete(num_states)
```

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+        if num_states is not None:
+            self.observation_space = Discrete(num_states)
+        if num_actions is not None:
+            self.action_space = Discrete(num_actions)
```
```suggestion
            self.action_space = spaces.Discrete(num_actions)
```

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+
+    def transition_fn(self, state, action):
+        """Sample next state according to T(s, a)."""
+        return sample_distribution(
```
```suggestion
        return util.sample_distribution(
```

AdamGleave commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+        self.seed()
+
+    def seed(self, seed=None):
+        """Set randomness seed."""
```
```suggestion
        """Set random seed."""
```

AdamGleave commented on src/seals/diagnostics/risky_path.py:
```diff
@@ -0,0 +1,58 @@
+    the risky path.
+    """
+
+    def __init__(self):
```
:+1: reward/transition matrix computation are clear

qxcv commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+            self.transition_matrix[state, action], random=self.np_random,
+        )
+
+    def ob_from_state(self, state):
```
Doesn't this only make sense for tabular envs, when the observation space is an int?

qxcv commented on src/seals/diagnostics/base_env.py:
```diff
@@ -0,0 +1,100 @@
+        state given by zero entries.
+        """
+        return np.zeros(
+            self.observation_space.shape, dtype=self.observation_space.dtype,
```
Consider adding an appropriate assertion, like `assert isinstance(self.observation_space, gym.spaces.Box)`, if you put this in a mixin that's likely to get used from another package.

qxcv commented on src/seals/diagnostics/risky_path.py:
```diff
@@ -0,0 +1,58 @@
+    """Environment with two paths to a goal: one safe and one risky.
+
+    Many LfH algorithms are derived from Maximum Entropy Inverse Reinforcement
+    Learning (Ziebart et. al, 2008), which models the demonstrator as producing
```
Where are (Ziebart, 2008) and (ZIebart, 2010) defined? He has a few papers from each of those years.

pedrofreire commented:
So, I propose then that we move `ResettableEnv` and `TabularModelEnv` to seals (I think seals could be seen as the LfH version of gym, which is gets used for its environment interface and utils, even when the environments themselves are not used).

I copied them from imitation and made some small modifications; I mostly just removed the abstract properties, and added the `transition_matrix` and `reward_matrix` as `__init__` arguments; I also replaced some abstract methods with reasonable default methods (most of the changes are in this diff: https://github.com/HumanCompatibleAI/seals/pull/20/commits/936e731235c3f0dfafe1013371c7aa1033757c78 ).

I also have to add some tests for the utils, to not lose test coverage.

qxcv commented on src/seals/diagnostics/__init__.py:
```diff
@@ -1,7 +1,9 @@
+gym.register(
+    id="seals/RiskyPath-v0",
+    entry_point="seals.diagnostics.risky_path:RiskyPathEnv",
+    max_episode_steps=5,
```
Seems easier to track it in one place. (the wrapper) rather than duplicating. Can't think of a strong reason why it _needs_ to be in the environment.

pedrofreire replied:
Yeah, I think the main use was to use the environment for tabular irl without needing to instantiate it through gym, but I guess you can just always use gym.
I also find slightly annoying that the horizon has to be placed in the gym.register, out of context of the rest of the environment definition, but that is not a big deal (we can always add a comment explaining the horizon)

pedrofreire commented on src/seals/testing/envs.py:
```diff
@@ -121,16 +121,19 @@ def test_seed(env: gym.Env, env_name: str, deterministic_envs: Iterable[str]) ->
 
     is_deterministic = matches_list(env_name, deterministic_envs)
-    same_obs = all(new_seed_equals_orig_rollout(seed) for seed in range(20))
+    same_obs = all(different_seeds_same_rollout(seed, seed + 1) for seed in range(100))
```
Yeah; the issue was that in RiskyPath, starting with action 0 leads to a deterministic path, while taking action 1 is stochastic; thus the test would fail half of the time.
More generally, there are environments that are only stochastic in some regions of the state/action space.
Varying the actions should be enough for now, though there could still be environments in the future in which there are stochastic regions, but you cannot reach them with a few random actions (which would cause the test to fail).

AdamGleave replied:
Oh, good catch.

pedrofreire commented on src/seals/base_envs.py:
```diff
@@ -0,0 +1,160 @@
+from seals import util
+
+
+class ResettableEnv(gym.Env, abc.ABC):
```
Hmm, I'm not sure what would be the best way of doing this - I guess you make a parameterized class or pass types in the constructor, but it is unclear for me whether pytype can use that - it would be better if we could infer types directly from the state and action spaces

AdamGleave commented on src/seals/base_envs.py:
```diff
@@ -0,0 +1,191 @@
+        return n_actions_taken >= self.horizon
+
+    @property
+    def feature_matrix(self):
```
Is the benefit of constructing this lazily that only some algorithms need it?
